orm/src/main/java/org/hibernate/search/batchindexing/impl/IdentifierProducer.java

```diff
@@ -93,21 +93,30 @@ public void run(StatelessSession upperSession) throws Exception {
 		log.trace( "finished" );
 	}
 
-	private void inTransactionWrapper(StatelessSession upperSession) throws Exception {
-		StatelessSession session = upperSession;
+	private void inTransactionWrapper(StatelessSessionImpl upperSession) throws Exception {
+		StatelessSessionImpl session = upperSession;
 		if ( upperSession == null ) {
 			if ( tenantId == null ) {
-				session = sessionFactory.openStatelessSession();
+				session = (StatelessSessionImpl) sessionFactory.openStatelessSession();
 			}
 			else {
-				session = sessionFactory.withStatelessOptions().tenantIdentifier( tenantId ).openStatelessSession();
+				session = (StatelessSessionImpl) sessionFactory.withStatelessOptions().tenantIdentifier( tenantId ).openStatelessSession();
 			}
 		}
 		try {
-			Transaction transaction = session.getTransaction();
-			transaction.begin();
-			loadAllIdentifiers( session );
-			transaction.commit();
+			Transaction transaction = session.accessTransaction();
+			final boolean controlTransactions = ! transaction.isActive();
```

Comment:
BTW this is pretty much what I proposed on the dev list, but there it's about making it a public API of Hibernate.
I'll not introduce such an helper here though to avoid reuse in a different context, as this specific check of `isActive()` isn't the universally correct strategy, it only is in this context as this block is wrapped by an `org.hibernate.search.batchindexing.impl.OptionallyWrapInJTATransaction` :-/

I'll agree that this detail doesn't look good, we'll come back on how to deal with CMT / JTA and all that. Will have to do for now though as I don't want to refactor this area heavily with 5.6 still in progress.